### DISCUSSION TITLE
Feature Request: Splitting up into smaller repositories

### DISCUSSION THREAD

**Discussion by L4stR1t3s** [o3de/o3de#4826] [Ideas]
**Is your feature request related to a problem? Please describe.**
Open3D Engine is a huge project, but most of it will be completely irrelevant to me (or anyone else looking to use it).
It takes space and time to build things you don't need.
Are there any plans to split up the project into smaller repositories? So we can more easily pick and choose what we want/need?

**Describe the solution you'd like**
Everything is located in a single large repository right now, while there is potential for splitting up Open3D Engine into smaller repositories.
If you are not going to use specific Gems for example, I don't see the need for having a local copy of them and building them. Due to their modular nature it should be possible to spread Gems across different repositories.
Files used by the tests could perhaps as well be made optional. If you are not going to change Open3D Engine code, but simply use it in a project, I see little use for running these tests anyway.

**Describe alternatives you've considered**
Stripping down the project yourself is possible, but it's quite some work, that has to be repeated every time you upgrade to a newer version.

**Additional context**
Being able to keep the size and build times as low as possible is an important factor when deciding on which third party library to use. Especially if you are going to be using it for a project with a lot of target platforms.
Keeping the amount of work needed to upgrade to a newer version low is also an important factor to consider.

**Comment by amznestebanpapp** [comment]
You can build pieces of it if you dont want all of it, it wont save space on source files, but it will save space on output.
You can e.g. build `AzCore` by just passing that target to build.

**Comment by byrcolin** [comment]
I agree with this sentiment which is why one of the main goals of sig-core is to identify and purpose RFC for removing any code/gem that should not be considered core, meaning needed to run the engine. If the intention here is the moving of all gems to different repos then I would disagree with that idea as some are undoubtedly core to the engine and should remain so, and would be unnecessarily modular as to not be in the main engine repo. I would be interested in which code/gems you had in mind so they can be brought up at the next sig-core meeting.

**Comment by lemonade-dm** [comment]
There are other options as well do reduce what even appears in the Engine build solution.
It is scriptable to be able to unregister gems with the engine, so that they aren't available to enabling/disabling gems(via o3de.py or ProjectManager) Unregistering a gem will also make it so that the gem doesn't appear in the build solution(ninja, Visual Studio, Xcode, etc..)

The engine.json contains a list of [external_subdirectories](https://github.com/o3de/o3de/blob/development/engine.json#L8-L88) which the engine CMake script hooks into to allow Gems that are distributed with the engine to be built.

That can be modified, via the o3de.py script to remove sets of gems that aren't desired to be part of an engine source "view"
For example running the following command will unregister the "ScriptCanvas" gem with the engine.
`scripts\o3de.bat register --gem-path Gems/ScriptCanvas --external-subdirectory-engine-path <engine-root> --remove`

**Comment by L4stR1t3s** [comment]
I think there is a huge difference in user friendliness of a system where you remove unnecessary parts, or where you add necessary parts.
A new user of Open 3D Engine will not know what is what, and will have no idea of what to strip. It will also be much harder to start using Open 3D Engine if the base repository is large with a lot of optional stuff. A new user doesn't know what is what, and the more stuff there is, the harder it becomes to find what you want/need.
A basic repository with only necessary stuff is much easier to start working with. It's also a lot easier to go and find stuff you want/need and add those, instead of removing what you don't need/want.

> I would be interested in which code/gems you had in mind so they can be brought up at the next sig-core meeting.

I don't really know. I only just started exploring Open 3D Engine, and well, I noticed that having everything thrown into a single repository doesn't exactly make it easier to start working with Open 3D Engine.

**Comment by AMZN-alexpete** [comment]
Though it's not easy to do, I think we should try to quantify what we mean by "large" using some metrics.  We need some kind of target so we know when we've satisfied making the engine not overly large.  

Currently, there are 3 driving factors (in my opinion)
1. Convenience for developers - splitting the code across more repositories makes large changes difficult and engine developers have to open multiple PRs, watch additional repositories and sync multiple repositories to their machine.  It also makes it, to a degree, more complex for our build machines to track and manage additional repositories.  However, for other developers - having multiple repositories is a good thing because it becomes easier to just get what you want.

2. Resource/cost constraints - there are limits to the GitHub free tier that we'd like to stay well under if possible, and limitations on the number and size of files you can have in a Git repository and it still perform well on an average developer machine.  This is one reason why we're using LFS for our large files.  Developers are also limited by their own hardware and networking constraints and we want to be mindful of that.  It may not seem like it, but we've actually massively brought our main repository size down and split up the repository to make it fit (some of our sample projects were 6+GBs)

3. Ownership - many awesome parts of the game engine (Kythera AI, PopcornFX, 3rdParty dependencies etc.) may need to come from outside the main repository because they are owned and operated by someone else that may not want their code to be in the main repository or organization for whatever reason (licensing, existing infrastructure etc).

Having a base repository that only has the minimum necessary code is possible, but we have to balance the benefits of a small main repository size with the cost to developers who would have to sync, maintain and work in additional repositories as a result.

**Comment by byrcolin** [comment]
I agree with @AMZN-alexpete , I think we should quantify things a bit more. And for this to be actionable we need a proposal for exactly what we are proposing to be removed, then we can dig into those specific codes/gems and come to a decision on specific items.
So I have started that: The engine looks to be around 10.8 gigs. 
<img width="713" alt="TopLevel" src="https://user-images.githubusercontent.com/68035668/126417014-4f498d2a-ca48-4ec2-94cd-3a42617413b1.PNG">
The largest contributor by far are Gems which account for 9.5gigs @ 87.9% of the total size of the repo. Of the Gems by far the largest are the 3 Atom gems which account for 67% of all gems.
<img width="712" alt="GemsTopLevel" src="https://user-images.githubusercontent.com/68035668/126417097-24834d16-15d0-4908-8352-b7bc8de4568c.PNG">
Within these two atom gems there appears to be a character "Lucy" which seems to be a wip (work in progress) So why is this in here? which comes in at a whopping 3.2gigs alone which is 99.7% of this gems size.
<img width="712" alt="Lucy" src="https://user-images.githubusercontent.com/68035668/126417329-1f8b16f8-7482-4346-ad65-7e9d5b141514.PNG">
In the other Atom Content gem is dominated by 8K textures which clock in between 100megs-1gig each which account for 99.3% of this gems size.
<img width="711" alt="Vegitation" src="https://user-images.githubusercontent.com/68035668/126417537-eeda9942-5d5b-40f3-b158-dc1718bfe4aa.PNG">
Vegetation gem is the next largest and has bunch of high fidelity rock textures and some vegetation which isn't so bad. The rest is relatively small potatoes.
If I remove just the lucy object and the 8K textures reduces the size of the repo to under 4.8 gigs almost a 50% reduction.
I propose we remove these items and make them optional gem downloads.
I would also add that the built versions of these source files are also nearly as big in the cache, so this would be a 2x or more is multiple asset platforms are enabled.

**Comment by L4stR1t3s** [comment]
What about the disk space required for a build? A build of all targets easily amounts to +100gb on Windows (no idea about other OS). Removing models, textures, ... will not decrease the space required for a build. It's currently not really feasible to build multiple build configurations on an average developer machine. Deleting and building different build configurations takes a lot of time.

I think the required disk space for a build should also be considered as a factor. I guess we'd need to collect data on a clean build of a specific configuration for several platforms, to get an overview.

**Comment by rgba16f** [comment]
I think it's also reasonable for us to pick some "average" spec machines and say the goal should be that the initial install build on those machines should be less than some duration? Does 30 Minutes seem like a reasonable duration for an Initial install build? 1 hour? 2 hours? 12 hours?
For machine specs I would go with either a base spec M1 Mac Mini, or a quad core windows pc with ssd.

**Comment by audiojak** [comment]
This is an interesting topic. It sounds like it could be even easier if we had use-case specific "distributions" of O3DE rather than just aiming for small core.

**Comment by amznestebanpapp** [comment]
> What about the disk space required for a build? A build of all targets easily amounts to +100gb on Windows (no idea about other OS). Removing models, textures, ... will not decrease the space required for a build. It's currently not really feasible to build multiple build configurations on an average developer machine. Deleting and building different build configurations takes a lot of time.
> 
> I think the required disk space for a build should also be considered as a factor. I guess we'd need to collect data on a clean build of a specific configuration for several platforms, to get an overview.

Sounds like you may be building everything. You dont need to do that...
CMake and MSBuild are target-based build systems. You can build just the targets you are interested in. For example, if you build the Editor, you will get just the enabled gems the Editor needs. Because the Editor depends on the AP, that will build it and also build the gems the AP needs.

**Comment by L4stR1t3s** [comment]
> 
> 
> > What about the disk space required for a build? A build of all targets easily amounts to +100gb on Windows (no idea about other OS). Removing models, textures, ... will not decrease the space required for a build. It's currently not really feasible to build multiple build configurations on an average developer machine. Deleting and building different build configurations takes a lot of time.
> > I think the required disk space for a build should also be considered as a factor. I guess we'd need to collect data on a clean build of a specific configuration for several platforms, to get an overview.
> 
> Sounds like you may be building everything. You dont need to do that...
> CMake and MSBuild are target-based build systems. You can build just the targets you are interested in. For example, if you build the Editor, you will get just the enabled gems the Editor needs. Because the Editor depends on the AP, that will build it and also build the gems the AP needs.

There are cases where building everything is necessary. Even when building everything in multiple configurations is needed. If you only touch a specific area of a project then you only need to build that specific area. But if you touch core areas, that affects most, if not all, of the project. Maybe I'm too cautious but I tend to make sure the areas my changes affect are at least checked if they compile.

**Comment by byrcolin** [comment]
@L4stR1t3s I agree that everything, build artifacts included, build times, should all be on the table for consideration. For me this all comes down to what we think the average game developer machine today looks like.
There is no real good source of data I'm aware of on what the average game development system is, average cpu, average gpu, average storage space, type of storage (io bandwidth), how much space it has devoted to development, etc... The steam survey is good data for average game player, not developer really... We should probably guess... So what should we assume the average game developer system looks like?
My 2 cents:
CPU: There are some really cheap cpus that have been out for a long time now that have 6-8 (or more) cores running around 3-4 GHz. So I think 6-8 cores at 3-4 GHz is a reasonable expectation for a developer.
Memory: 16 gigs seems reasonable today. (You are going to need more if you are going to link in release with all the optimizations turned on, 32-64 gigs would be preferred there, as release optimizations take a lot of memory.)
Storage: I think its pretty safe to assume the user will have at least @ 1TB drive on average. The 1TB ssd, and increasingly nvme ssd, seems to be the sweet spot. If they are going to do game development, that should be enough for a small project. And if you're doing multiple small game development, or a medium size project, or multi platform single small project I don't think its unreasonable for the game dev to expect more drive space will be required, you are now doing some significant game development at this point after all. So maybe multiple 1TB ssds, one per project/platform should not be unreasonable expectation here. We all had a 2TB ssd on the last commercial game I made with the engine, it was multi-platform and it was enough (though barely). I don't think you'll get any disagreement that a smaller space requirement is desirable, however we would need a proposal for how do we would go about achieving that.
GPU: I think a 1070 level graphics can be expected now. Its 2 iterations old now, pretty average these days.
Build times: This is going to highly depend on the combination of hardware, the engine type (are you using the prebuilt engine or building everything yourself), project size, configuration you are building and options selected... so I think time in minutes is kind of red herring, to many factors... If we are talking about the theoretical average hardware, if we assume we are building a standard downloadable tutorial type game project (not automatedtesting that ships with the engine, its not really indicative of a real small game), if we assume you are building the engine yourself and not using the pre-built engine, if we assume you are building monolithically, if we assume you are using unity building options, if we assume you are building from 0, then perhaps we could get to a minutes estimate... With all those conditions, I would put forward 45 minutes as a reasonable expectation. However, I do not think this is indicative of the average game developer, perhaps it is of a build server or lead programmer making a build for the team, but the team members would then most likely be use that build to iterate on the game. Artists, level designers, game play people, QA, will mostly downloading pre-built code/assets. That iteration time is going to be WAY lower. If I was to use this theoretical system today with the repo as is (maybe getting rid of the Lucy and 8K assets) I think we are probably in the 45 minute neighborhood.
Lowering the storage space requirement is largely a function of assets at this point. So elimination of extremely large assets and increased modularity (removing from core engine into downloadable gems) is probably the way to most effectively reduce the core install size and thus minimum build size for code and assets. Lowering the storage requirement for the non prebuilt engine without losing the interim build artifacts is going to be harder to do. I don't currently have a good proposal for doing so, do you or anyone else have a ideas/a proposal for lowering the full build space, requirements?

**Comment by amznestebanpapp** [comment]
> There are cases where building everything is necessary. Even when building everything in multiple configurations is needed. If you only touch a specific area of a project then you only need to build that specific area. But if you touch core areas, that affects most, if not all, of the project. Maybe I'm too cautious but I tend to make sure the areas my changes affect are at least checked if they compile.

Even if you touch core areas you dont need to build everything. The build system is a DAG of targets and dependencies. At the end of that DAG, there are leaf targets. Building "all" means that every leaf target is built, that includes tests, tools, every exe/app/etc. Most of the targets are usually not used by the Editor or a game project. 

Gems are modules that can be enabled or not in a project. When a gem is enabled in a project, that gem is added as a dependency in that DAG. Therefore, if your project depends on e.g. the EMotionFX Gem, when building your GameLauncher or the Editor, EMotionFX gets built. However, if your project does not depend on e.g. the Microphone gem, then that doesnt get built (when building the Editor).

The point I am trying to make is that if you are trying to reduce what you are using of the engine, you can reduce what you are building by building the specific leaf targets that are of your interest. 

Previously, in Lumberyard, we had a big issue with this. Gems were enabled or not as targets depending if your project had them enabled or not. That produced issues with gems never being built by CI. Some developers just added them to the test projects so they would get built (despite the gem not being used by the test project at all). Some gems were not being built at all.
Thats why we changed the approach and have them now all enabled as targets, so we can validate all of them build and all the tests get run. That doesnt mean that a developer needs to continuously be building those.

My workflow is usually iterating on specific targets where I am making modifications, then I go to e.g. the Editor target and run that to validate it in the Editor, then go to the GameLauncher and run that to validate it in the Game. Depending on the change, I do a full build, just to reduce burden on CI. However, CI runs an all build, so even if your change breaks some code that is not being considered by the leaf targets that you are picking, it will be caught by CI.

Building all in some cases is inevitable. Even if things are in different repositories, you would have to pull all the repositories downstream and validate that your change doesnt break them. And it will make things more complicated because you now need to coordinate that change across multiple repositories. If it is a non-backwards compatible change, that means that you will end up with a window where you are going to break all downstream repositories (either you merge first the core change or merge each repository, in both cases the code is not in sync for a time window). It could even be impossible to submit the change since CI needs to be green for your change to go through.

**Comment by L4stR1t3s** [comment]
Does anyone have experience with git submodules? https://gist.github.com/gitaarik/8735255

It seems like it might be able to help with keeping track of commits to multiple repositories.

There could be a couple of repositories like o3de-core, o3de-tools, o3de-gems (not suggesting these as actual repositories, just an example). Users could then start with o3de-core and add repositories as needed/wanted.
There could then also be an over-arching o3de-dev that contains all those repositories as submodules.

I might play around with git submodules a bit, first time I come across them and it looks interesting.

**Comment by L4stR1t3s** [comment]
@amznestebanpapp

You approach this with the assumption that everyone knows as much about the Open3D Engine project as you. Building only what you need sounds good, but how does someone who has never used Open3D Engine before know what that is?

**Comment by rgba16f** [comment]
We did try git submodules before moving to github. The experience was bad enough that we abandoned them for github :).

**Comment by L4stR1t3s** [comment]
> I don't currently have a good proposal for doing so, do you or anyone else have a ideas/a proposal for lowering the full build space, requirements?

Significantly lowering the full build space requirements can only be done by lowering the amount of code. What code can be considered to be not core and therefor be moved into another repository? Some of the gems? I haven't really dug deep into them at all yet, so I'm not ideally placed to make any decisions.
Are all of the tools considered to be "core"? Or would it be feasible to make some of them optional? I haven't looked into what they all do, so I don't really know how essential they are.

I've seen comments on the Discord about Gems using third party solutions that require a license. I think those would definitely be good candidates to move into a separate repository. Gems that require a user to get a license with a third party can hardly be considered to be "core" of an open source project.

**Comment by lemonade-dm** [comment]
> We did try git submodules before moving to github. The experience was bad enough that we abandoned them for github :).

To clarify, the issues with git submodules were that any changes that occurs across repos requires two PRs and can not be committed atomically. 
i.e If there is a change in the AzCore library, that facilitated a change in the ScriptCanvas Gem, then the change to the AzCore would first have to be accepted to the Core Repo, followed by updating the submodule link in the Gem Repo as well as the ScriptCanvas code. Afterwards the ScriptCanvas changes is then Pull Requested to add to Gem Repo.

It also caused developers issues syncing the correct commit from the submodule when updating the O3DE engine. Syncing the latest commit of the Core Repo causes build issues if the O3DE Repo is not using that commit.
But on the other hand if the developer wanted to update files within the Core Repo, they would need to make sure they are working on the latest commit.

It was quite easy for a developer to make a change in the submodule based on an older commit and push that change up for PR.
In that case it was hard for a developer to tell which of the newer commits in the submodule repo, that wasn't theirs should be added to the new submodule link

**Comment by L4stR1t3s** [comment]
> 
> 
> > We did try git submodules before moving to github. The experience was bad enough that we abandoned them for github :).
> 
> To clarify, the issues with git submodules were that any changes that occurs across repos requires two PRs and can not be committed atomically.
> i.e If there is a change in the AzCore library, that facilitated a change in the ScriptCanvas Gem, then the change to the AzCore would first have to be accepted to the Core Repo, followed by updating the submodule link in the Gem Repo as well as the ScriptCanvas code. Afterwards the ScriptCanvas changes is then Pull Requested to add to Gem Repo.
> 
> It also caused developers issues syncing the correct commit from the submodule when updating the O3DE engine. Syncing the latest commit of the Core Repo causes build issues if the O3DE Repo is not using that commit.
> But on the other hand if the developer wanted to update files within the Core Repo, they would need to make sure they are working on the latest commit.
> 
> It was quite easy for a developer to make a change in the submodule based on an older commit and push that change up for PR.
> In that case it was hard for a developer to tell which of the newer commits in the submodule repo, that wasn't theirs should be added to the new submodule link

Well, it did sound too good to be true...

**Comment by amznestebanpapp** [comment]
> @amznestebanpapp
> 
> You approach this with the assumption that everyone knows as much about the Open3D Engine project as you. Building only what you need sounds good, but how does someone who has never used Open3D Engine before know what that is?

Maybe it came through the wrong way, it wasn't the intention. I was trying to provide other ways to address the issue and explain the hows and whys. Please dont take this as me pushing back on your request. Your request is valid, I am trying to understand the root causes of it.
This request could end up in a need to update documentation for others to be aware and improve their workflows or it could end up in a considerable effort.

The build instructions don't instruct to build all, they instruct to build particular targets: e.g. `AutomatedTesting.GameLauncher AssetProcessor Editor`. That was on purpose. So maybe we need a section explaining how to reduce the scope of your build by picking targets or make more emphasis on why building "all" is not the expected workflow. Would that cover your concerns/issues?

I am still trying to understand the root causes and what the best solution is, Despite it sounds like a good idea to cut the project in half or in N parts, its a trade off, it brings all sorts of other problems. Currently we have ~3M of C++ code, its a big project, there is some cleanup that can be done, but seems your expectation is that it will be 1/10 or 1/20 of the size.
Practically, it would take a huge amount of work to achieve that (if it is even possible). 

There has been work to improve build times and reduce code size, more than 1.5M LOCs of code were removed from Lumberyard and tenths of Gb were removed from the repository. Build times were also reduced considerably, in comparison, O3DE takes ~1/3 of the time to build that LY did. 
Just want to show that there has been work there and we will continue improving it. The stats https://github.com/o3de/o3de/graphs/code-frequency also show a good frequency of removals. 

This is also not a particular thing of O3DE, is how most build systems and C++ projects work. It gets particularly complicated in this scenario because of the size of the project and amount of features available. 

Lets keep exploring on the issues and options to solve it, by no means is my intention to discourage that.
So far, what I have collected is:
1) long build times
2) build output takes a lot of space
3) project takes a lot of space

As others stated, we need some numerical definitions to get a better sense of expectation. Some may consider 100Gb acceptable, some may not. 

For (1) and (2) a partial solution is to build less using the method I explained before. We could document this better.
For (3) we could cleanup some, but we can also maybe do git sparse checkout and filter out gems that occupy a lot of space and are not going to be used. We could built a tool to do this or document how to do it.

**Comment by L4stR1t3s** [comment]
> This request could end up in a need to update documentation for others to be aware and improve their workflows or it could end up in a considerable effort.

> The build instructions don't instruct to build all, they instruct to build particular targets: e.g. `AutomatedTesting.GameLauncher AssetProcessor Editor`. That was on purpose. So maybe we need a section explaining how to reduce the scope of your build by picking targets or make more emphasis on why building "all" is not the expected workflow. Would that cover your concerns/issues?

That would help alleviate the new user finding his way issue.

> but seems your expectation is that it will be 1/10 or 1/20 of the size.

I am definitely not expecting that much. At most 1/2 and I would consider that a lot already. I'd be happy with 3/4 or 4/5.

> Lets keep exploring on the issues and options to solve it, by no means is my intention to discourage that.
> So far, what I have collected is:
> 
>     1. long build times
> 
>     2. build output takes a lot of space
> 
>     3. project takes a lot of space

4. Git works better with smaller repositories. Especially in the long run.

5. Easier to add optional stuff, than to remove it (especially for new users).

6. For me personally it's also about how Open3D Engine could be different from other existing projects like it. Personally I don't like any of the other ones, because they force you into a specific workflow. And in my opinion this happens because they aren't really modular. They claim to be, and you could indeed replace most of their systems with your own, but it's a very difficult task because everything is so interconnected. Real modularity can only be achieved if the core developers really treat it like a modular project. From what I've seen so far Open3D Engine can achieve the level of modularity I'd like to have. I actually recently decided I wanted to take a look at Lumberyard, and it seems I was a little late for LY, but perfectedly on time for O3DE.
